### DISCUSSION TITLE
Mosh-server issue after Mac update to Sequoia

### DISCUSSION THREAD

**Discussion by juanpark** [blinksh/blink#2054] [Support]
After updating to iOS 18 on my iPad and Sequoia on my Mac, mosh-server no longer seems to be working on the Mac. Mosh from Blink to other linux devices work fine. Mosh-server on Mac is not working with the following error:

dyld[43782]: Library not loaded: /opt/homebrew/opt/protobuf/lib/libprotobuf.27.1.0.dylib
Referenced from: <55EBD7C0-6AE1-3B19-8CAD-A48819258BB5> /opt/homebrew/Cellar/mosh/1.4.0_16/bin/mosh-server
Reason: tried: '/opt/homebrew/opt/protobuf/lib/libprotobuf.27.1.0.dylib' (no such file), '/System/Volumes/Preboot/Cryptexes/OS/opt/homebrew/opt/protobuf/lib/libprotobuf.27.1.0.dylib' (no such file), '/opt/homebrew/opt/protobuf/lib/libprotobuf.27.1.0.dylib' (no such file), '/opt/homebrew/Cellar/protobuf/28.1/lib/libprotobuf.27.1.0.dylib' (no such file), '/System/Volumes/Preboot/Cryptexes/OS/opt/homebrew/Cellar/protobuf/28.1/lib/libprotobuf.27.1.0.dylib' (no such file), '/opt/homebrew/Cellar/protobuf/28.1/lib/libprotobuf.27.1.0.dylib' (no such file)

When trying to reinstall protobuf:

checking whether protoc matches protobuf... no
In file included from conftest.cpp:68:
./conftest.pb.h:9:10: fatal error: 'limits' file not found
9 | #include
| ^~~~~~~~
1 error generated.
configure: error: in `/private/tmp/mosh-20240925-44975-vmkd63/mosh-1.4.0':
configure: error: Could not build output generated by protoc (/opt/homebrew/opt/protobuf/bin/protoc).
Check that protoc matches the installed protobuf headers and libraries.

Anybody having similar issues? Any solutions?

**Comment by carloscabanero** [comment]
This looks like an issue on Homebrew, so you may want to look there. But, the simple solution with Blink is to install a static mosh we provide with the `—install-static` flag.

**Reply by juanpark** [reply]
Thank you for your suggestion. But the problem persists:

Per you suggestion when I run the following on Blink on my iPad:
$mosh --install-static mac_server

I get back:
"mosh: Nothing received from server on UDP port 60001."

I've check to see if the port is open with nc and it is. Also my router's port forwarding rules are properly set, which I don't think is an issue as the same failure happens from within the network.

When I turn off the firewall on my mac, it hangs indefinitely. When I toggle the firewall on my mac, it hangs, but for some reason the mac terminal prompt become visible, though it isn't responding with an error message on top saying "mosh: Last contact xx:xx ago."


FYI: mosh-server seems to be back in some semblance with the following message but still can't connect to it with the same message of "mosh: Nothing received from server on UDP port 60001."

MOSH CONNECT 60001 hH0CWVIE5YiJm+FdBM6/zA

mosh-server (mosh 1.4.0) [build mosh 1.4.0]
Copyright 2012 Keith Winstein <mosh-devel@mit.edu>
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>.
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.

[mosh-server detached, pid = 45395]

**Reply by carloscabanero** [reply]
The problem doesn’t persist, it is a different problem, different is good 😉

mosh-server is now running and we get the connection info but your client cannot connect to it. Have you tried https://github.com/blinksh/blink/discussions/1932?

**Reply by juanpark** [reply]
Yes, different is good!

I've actually used the following script to get mosh to work on the mac with no issues until the upgrade to Sequoia:

#!/bin/sh
## Setup
COMMAND="sudo /usr/libexec/ApplicationFirewall/socketfilterfw"
MOSH=`which mosh-server`
REALMOSH=`readlink -f $MOSH`
FWSTATE=`$COMMAND --getglobalstate`
MOSHCHECK="$COMMAND --listapps"

## Is it in the firewall?
if $MOSHCHECK | grep -q $REALMOSH; then
	echo "$REALMOSH is present."
else
	echo "$REALMOSH is not present. Adding."
        $COMMAND --add $MOSH
fi

## Unblock it
$COMMAND --unblockapp $REALMOSH

## Toggle firewall if it's enabled
case "$FWSTATE"
in
  *enabled*)  echo "Firewall Enabled. Toggling state."
        $COMMAND --setglobalstate off
        $COMMAND --setglobalstate on
     ;;
  *disabled*)       echo "Firewall Disabled. Not enabling."
     ;;
   *) echo "Unknown response for firewall state."
esac

Since this is no longer working, I did try out the script from #1932, which as can be seen is the same idea as the previous script , and this also didn't work.

As for trying out Tailscale, I am already using mosh with Zerotier as my mac's ssh is not exposed to the outside. I'm not sure how this would be different, and if this is just trying to skirt the underlying firewall (?) issue.

**Reply by carloscabanero** [reply]
Thanks for the thorough reply! I know I don't have problems with my main machine but I will try to play a bit. Haven't heard anybody else having the same issue, so that's why I'm thinking maybe it is some change in the firewall configuration that is not playing well with your configuration? Let's see if other people jump in.

**Reply by juanpark** [reply]
Thank you for your time. I will also try tweaking settings further and post anything that might shed light on the issue.

**Reply by carloscabanero** [reply]
So it looks like you are definitely onto something here. I'm able to replicate this in a clean installation of Sequoia. `mosh-server` starts, so I'm going to discard issues with sandboxing (https://troz.net/post/2024/sequoia_app_permissions/ via Discord). I do have Firewalls disabled and I was able to start a netcat server-client over UDP without issue. There are some reports of UDP traffic having issues with VPNs at the moment, so I wonder if the mosh server is somehow affected by these changes?

In any case, I think this would be something we may want to raise to upstream as the regular client-server is affected too.

**Reply by carloscabanero** [reply]
I did a few other quick tests. So when running client-server, the client is not receiving back from mosh-server, but the server is receiving from the client. I traced a running mosh-server and started typing on the "non-receiving client":

```
 ~ % sudo dtruss -p 940 -t write
SYSCALL(args) 		 = return
write(0x7, "j\0", 0x1)		 = 1 0
write(0x7, "h\0", 0x1)		 = 1 0
write(0x7, "fg\0", 0x2)		 = 2 0
write(0x7, "j\0", 0x1)		 = 1 0
write(0x7, "g\0", 0x1)		 = 1 0
write(0x7, "l\0", 0x1)		 = 1 0
write(0x7, "k\0", 0x1)		 = 1 0
...
write(0x7, "\r\0", 0x1)		 = 1 0
write(0x7, "l\006\0", 0x1)		 = 1 0
write(0x7, "kj\0", 0x2)		 = 2 0
write(0x7, "lkj\0", 0x3)		 = 3 0
write(0x7, "lkj\0", 0x3)		 = 3 0
```

Only traffic from iPad to MBA
```
 ~ % sudo tcpdump port 60001
tcpdump: data link type PKTAP
tcpdump: verbose output suppressed, use -v[v]... for full protocol decode
listening on pktap, link-type PKTAP (Apple DLT_PKTAP), snapshot length 524288 bytes
15:57:46.740094 IP ipad.50061 > macbookair.60001: UDP, length 74
15:57:49.708756 IP ipad.50061 > macbookair.60001: UDP, length 77
15:57:52.780661 IP ipad.62441 > macbookair.60001: UDP, length 64
15:57:54.727691 IP ipad.62441 > macbookair.60001: UDP, length 83
15:57:54.727697 IP ipad.62441 > macbookair.60001: UDP, length 74
```
(A healthy connection would send the heartbeat both ways, but here we only have the client to server)

**Reply by carloscabanero** [reply]
@juanpark I can confirm things work with Tailscale. It looks like VPNs are having some type of issue, and maybe your Zerotier is being hit by the same firewall issue as mosh is.

**Reply by juanpark** [reply]
Thank you for looking into this. 

**My Tests:
Mosh works from Blink to:**
- my Raspberry Pi machine over ZeroTier.
- my Windows 10 machine over ZeroTier.

**Mosh still doesn't work from Blink to:**
- my Mac Mini (Sequoia) over ZeroTier. (ssh works fine over ZeroTier)
- my Mac Mini (Sequoia) not over ZeroTier (internal network). 
- my Macbook (Catalina) over ZeroTier. (ssh works fine over ZeroTier)
- Couldn't test as no way to reach it from outside other than over ZeroTier (internal network at work).

Seeing that my Macbook with Catalina is having the same issue with Mosh that my machine with Sequoia is having makes me feel that perhaps the issue is not with the update. But what has me confused is that the problem is limited to only my Mac machines and ZeroTier doesn't seem to be the issue as 1. my mosh connections to other machines over ZeroTier work with no issue, and 2. my attempts to mosh into my Mac machines without ZeroTier have the same issues.

**Reply by carloscabanero** [reply]
It is looking like an issue with the firewall on the Mac. But I couldn't find any incantation that fixes it. And from my tests, it looks like it is blocking the "outbound" traffic, which is even weirder. Will continue playing with it, but we may need to reach the dev forums then.

**Reply by carloscabanero** [reply]
We found what the issue is (many thanks to rrgeorge), but don't have an easy solution for it yet. The good news is that on your side, it looks like it should be ZeroTier the one solving it too, and once they do, you should be able to run mosh-server normally:  https://github.com/mobile-shell/mosh/issues/1254#issuecomment-2405102497

**Reply by juanpark** [reply]
That is very good news. Thank you for the effort. I will be patiently waiting for the update.

**Reply by fmt** [reply]
@carloscabanero Any news on this issue?

**Reply by carloscabanero** [reply]
Starting from the workaround. As this is an issue with protections over the local network, the easy fix is to connect over a network like Tailscale. Then Tailscale is responsible to request access and all connections made through it will go through.

What we found is that due to the way the Local Network protection works, there always needs to be a parent process hanging that is responsible for the Local Network permission. The main problem with fixing mosh-server on its own is the ssh -> mosh-server flow. SSH doesn't need any special network permissions, but when it disconnects after starting mosh-server, mosh-server will be "orphaned".

We tried a variety of things like trying to have mosh-server request permission on its own, etc... but none of them are trivial. The new macOS doesn't have any new ways to handle this (like manually adding a command to the "firewall" list), and from what I discussed with engineers at WWDC, they were not sure that would be something they wanted to add.

Another workaround is to start mosh as an agent somehow. We are considering something like this for a Blink version on the Mac. But this is all on paper at the moment.
